README.md

```diff
@@ -0,0 +1,188 @@
+
+# Sample Android Native Application
+
+## Introduction
+
+This is a sample Android Java application showing use of DataSync, [Keycloak](https://www.keycloak.org/about.html) and Unified Push using native upstream SDK's.
+Application is sending requests to [Ionic showcase server]([https://github.com/aerogear/ionic-showcase/tree/master/server](https://github.com/aerogear/ionic-showcase/tree/master/server)) which is a GraphQL server. For DataSync, application uses Apollo Client to query, mutate and subscribe. [AppAuth](https://github.com/openid/AppAuth-Android) to connect
+ with Keycloak and Aerogear SDK for Unified Push support.
+
+## Implementation
+### 1. DataSync
+#### Generating queries, mutations and subscriptions
+  To generate queries, mutations and subscriptions of running GraphQL server [Apollo Codegen]([https://github.com/apollographql/apollo-tooling](https://github.com/apollographql/apollo-tooling)) was used.
+
+#### Creating client
+ - First, we need to build OkHttpClient to handle network requests
+  ```java
+OkHttpClient okHttpClient = new OkHttpClient
+        .Builder()
+        .addInterceptor(chain -> {
+            Request original = chain.request();
+  Request.Builder builder = original.newBuilder().method(original.method(), original.body());
+  builder.header("Authorization", authHeader);
+ return chain.proceed(builder.build());
+  })
+        .build();
+```
+ - Then, NormalizedCacheFactory to provide caching mechanism
+ ```java
+ ApolloSqlHelper apolloSqlHelper = new ApolloSqlHelper(context, DB_CACHE_NAME);
+
+NormalizedCacheFactory<LruNormalizedCache> cacheFactory = new LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION)
+        .chain(new SqlNormalizedCacheFactory(apolloSqlHelper));
+ ```
+ - Passing in connection params for subscriptions to work if server is behind authentication mechanism
+ ```java
+ Map<String, Object> connectionParams = new HashMap<>();
+connectionParams.put("Authorization", authHeader);
+```
+- And build our Apollo Client
+```java
+return ApolloClient.builder()
+        .enableAutoPersistedQueries(true)
+        .serverUrl(serverUrl)
+        .normalizedCache(cacheFactory, resolver)
+        .subscriptionConnectionParams(connectionParams)
+        .subscriptionTransportFactory(new WebSocketSubscriptionTransport.Factory(serverUrl, okHttpClient))
+        .okHttpClient(okHttpClient)
+        .build();
+```
+#### Using queries, mutation and subscriptions
+Once client is build we can use it to run queries, mutations and subscriptions. On application start, a query, mutation or subscription is build
+#### Query
+```java
+MyGeneratedQuery nameOfMyQuery = MyGeneratedQuery
+        .builder()
+        .build();
```

Comment:
where is this coming from? Is it part of the app, or just an example for this doc?

Reply:
this is just an example for the doc. Every builder whether it is query mutation or subscription looks the same so I thought it would make sense to make it a little bit more generic.

Reply:
I had changed it to match app code. Very good point, I don;t think it actually makes any sense to have one of 3 code snippets trying to be generic while others are just matching the code itself.